In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

import os
print(os.listdir("../input"))

# Any results you write to the current directory are saved as output.

['train.csv', 'sample_submission.csv', 'test.csv']


In [2]:
train = pd.read_csv("../input/train.csv")
train.head(15)

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0       1       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       1       0       0       0       0       0       0       0       0   
3       4       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   
5       0       0       0       0       0       0       0       0       0   
6       7       0       0       0       0       0       0       0       0   
7       3       0       0       0       0       0       0       0       0   
8       5       0       0       0       0       0       0       0       0   
9       3       0       0       0       0       0       0       0       0   
10      8       0       0       0       0       0       0       0       0   
11      9       0       0       0       0       0       0       0       0   
12      1       0       0       0       0       0       0       0       0   
13      3       0       0       0       0       0       0       0       0   
14      3       0       0       0       0       0       0       0       0   

    pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0        0  ...         0         0         0         0         0         0   
1        0  ...         0         0         0         0         0         0   
2        0  ...         0         0         0         0         0         0   
3        0  ...         0         0         0         0         0         0   
4        0  ...         0         0         0         0         0         0   
5        0  ...         0         0         0         0         0         0   
6        0  ...         0         0         0         0         0         0   
7        0  ...         0         0         0         0         0         0   
8        0  ...         0         0         0         0         0         0   
9        0  ...         0         0         0         0         0         0   
10       0  ...         0         0         0         0         0         0   
11       0  ...         0         0         0         0         0         0   
12       0  ...         0         0         0         0         0         0   
13       0  ...         0         0         0         0         0         0   
14       0  ...         0         0         0         0         0         0   

    pixel780  pixel781  pixel782  pixel783  
0          0         0         0         0  
1          0         0         0         0  
2          0         0         0         0  
3          0         0         0         0  
4          0         0         0         0  
5          0         0         0         0  
6          0         0         0         0  
7          0         0         0         0  
8          0         0         0         0  
9          0         0         0         0  
10         0         0         0         0  
11         0         0         0         0  
12         0         0         0         0  
13         0         0         0         0  
14         0         0         0         0  

[15 rows x 785 columns]

Data description:
The data files train.csv and test.csv contain gray-scale images of hand-drawn digits, from zero through nine.

Each image is 28 pixels in height and 28 pixels in width, for a total of 784 pixels in total. Each pixel has a single pixel-value associated with it, indicating the lightness or darkness of that pixel, with higher numbers meaning darker. This pixel-value is an integer between 0 and 255, inclusive.

The training data set, (train.csv), has 785 columns. The first column, called "label", is the digit that was drawn by the user. The rest of the columns contain the pixel-values of the associated image.

Each row represents the pixel intensity matrix of an image

In [ ]:
train.iloc[512,:]

In [ ]:
train.loc[train.loc[:,'label']==2,:]

Beautiful way of accessing dataframes : loc (label based) and iloc(position based)

read more here: https://brohrer.github.io/dataframe_indexing.html

In [ ]:
from sklearn.model_selection import train_test_split

x=np.array(train.drop('label',axis=1).values)
#it does not drop it permanently from the database, to drop permanently use inplace=true arguement
y=np.array(train.loc[:,'label'])

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

Learn(t) how to drop columns in pandas and how to do it temporarily. converting dataframes to array. and splitting into a train test set

In [ ]:
from matplotlib import pyplot as plt
plt.imshow(x[6,:].reshape(28,28), cmap='gray',interpolation='none')
plt.axis('off')
plt.show()

In [ ]:
def display(x,y=-1) :
    plt.imshow(x.reshape(28,28), cmap='gray',interpolation='none')
    plt.axis('off')
    if(y!=-1): 
        plt.title(y)
    plt.show()

In [ ]:
display(x[6],y[6])

In [ ]:
for i in range(15):
    display(x[i],y[i])

# Decision Tree Classifier

In [ ]:
from sklearn.tree import DecisionTreeClassifier

decisionTreeModel=DecisionTreeClassifier().fit(x_train,y_train)
modelPrediction=decisionTreeModel.predict(x_test)
import sklearn
sklearn.metrics.accuracy_score(y_test, modelPrediction) 

In [ ]:
def accuracy(y_test,y_pred):
    r=sklearn.metrics.accuracy_score(y_test, y_pred)*100 
    print('The accuracy score is '+str(r)+' percent')

In [ ]:
accuracy(y_test,modelPrediction)

In [ ]:
from sklearn.metrics import confusion_matrix 
cm = confusion_matrix(y_test, modelPrediction)
print(cm)

In [ ]:
import seaborn as sns
sns.heatmap(cm,annot=True,cmap="YlGnBu")

# Random Forest Classifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier(n_estimators=150)
clf.fit(x_train,y_train)
forestPredictions=clf.predict(x_test)
sklearn.metrics.accuracy_score(y_test,forestPredictions)

In [ ]:
from sklearn.metrics import confusion_matrix 
cmf = confusion_matrix(y_test, forestPredictions)
sns.heatmap(cmf,annot=True,cmap="YlGnBu")

# SVM classifier

In [ ]:
#Don't run it without normalisation

#from sklearn.svm import SVC 
#svm_model_rbf = SVC(kernel = 'rbf').fit(x_train, y_train) 
#svm_predictions = svm_model_rbf.predict(x_test) 
#sklearn.metrics.accuracy_score(y_test,svm_predictions)

## Results after normalisation

In [ ]:
x_test=x_test/255
x_train=x_train/255

In [ ]:
from pprint import pprint
pprint(x_train[3].reshape(28,28))

In [ ]:
from sklearn.tree import DecisionTreeClassifier

decisionTreeModel=DecisionTreeClassifier().fit(x_train,y_train)
modelPrediction=decisionTreeModel.predict(x_test)
import sklearn
sklearn.metrics.accuracy_score(y_test, modelPrediction) 

In [ ]:
clf=RandomForestClassifier(n_estimators=150)
clf.fit(x_train,y_train)
forestPredictions=clf.predict(x_test)
sklearn.metrics.accuracy_score(y_test,forestPredictions)

# Real SVM (SVM didn't work without normalisation)

In [ ]:
svm_model_rbf = SVC(kernel = 'rbf').fit(x_train, y_train) 
svm_predictions = svm_model_rbf.predict(x_test) 
sklearn.metrics.accuracy_score(y_test,svm_predictions)

# Logistic Regression


In [ ]:
from sklearn.linear_model import LogisticRegression

LogisticRegressionModel=LogisticRegression().fit(x_train,y_train)
LogisticRegressionPrediction=LogisticRegressionModel.predict(x_test)
sklearn.metrics.accuracy_score(y_test,LogisticRegressionPrediction)

In [ ]:
cml = confusion_matrix(y_test, LogisticRegressionPrediction)
sns.heatmap(cml,annot=True,cmap="YlGnBu")